design/worker-pods.md

```diff
@@ -0,0 +1,161 @@
+
+### Open items
+
+- It probably makes sense to use Kubernetes Jobs to control worker pods, rather than directly creating "bare pods". The Job will ensure that a worker pod successfully runs to completion.
+- Currently, restic repository lock management is handled by an in-process lock manager in the Velero server. In order for backups/restores to safely run concurrently, the design for restic lock management needs to change. There is an [open issue](https://github.com/heptio/velero/issues/1540) for this which is currently scoped for the v1.1 release.
+- There are several prometheus metrics that are emitted as part of the backup process. Since backups will no longer be running in the Velero server, we need to find a way to expose those values. One option is to store any value that would feed a metric as a field on the backup's `status`, and to have the Velero server scrape values from there for completed backups.
+- Over time, many completed worker pods will exist in the `velero` namespace. We need to consider whether this poses any issue and whether we should garbage-collect them more aggressively.
+
+## Alternatives Considered
+
+### Enable >1 worker per controller
+
+One alternative option is to simply increase the number of workers being used per controller. This would, in theory, allow multiple backups/restores to be processed concurrently by the Velero server. The downsides to this option are that (1) this doesn't do anything to improve Velero's scalability, since all of the work is still being done within a single pod/process, and (2) it doesn't easily enable log streaming for in-progress backups.
+
+### Run multiple replicas of the Velero server
+
+Another option is to increase the number of replicas in the Velero deployment, so that multiple instances of the Velero server pod are running. This would require adding some sort of "leader-election" process for each backup to ensure that only one replica was processing it. It also would result in multiple replicas of the non-core controllers (backup deletion controller, GC controller, download request controller, etc.), which would require modifications to ensure there wasn't any duplicate processing. Additionally, this approach also doesn't easily enable log streaming for in-progress backups.
+
+## Security Considerations
+
+Each worker pod will be created in the `velero` namespace, will use the `velero` service account for identity, and will have the `cloud-credentials` secret mounted as a volume. As a result the worker pods will have the same level of privilege as the Velero server pod.
```

Comment:
Presumably we could lower the privileges of the Velero server if it now only needs to create these Jobs/Pods and no longer needs cluster-admin access for backing up all the resources.

Reply:
The Velero server would be reduced, but Velero overall would still need the cluster-admin access.

So it does change, but I think in terms of the overall set of permissions, it doesn't change things.

That said, this design _may_ enable the server to instantiate a worker pod with the requesting user's same permissions, and the backup flow within that worker pod could differentiate between not found and permission denied errors to handle logging. This requires more detailed thought, though, as even then access to cluster resources is needed if we continue to do things like follow a Pod->PV->PVC.

Reply:
Yeah you're right that the workers will still need cluster-admin, though this still feels like an improvement as the long-running server process no longer needs it.

> That said, this design may enable the server to instantiate a worker pod with the requesting user's same permissions

Interesting, how do you see this working? Would the user need to specify the ServiceAccount to use for the worker?

Reply:
That's a good question. It could be exposed for them to set, or maybe it's created on the fly to match the current user? That may be tricky to get right, so likely less desirable.